### PR TITLE
New debug adapter

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

### CODE REVIEW

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ProtocolDispatcher.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    @Override
+    public Response dispatchRequest(Request request) {
+        Messages.Response response = new Messages.Response();
+        response.request_seq = request.seq;
+        response.command = request.command;
+        response.success = true;
+
+        Command command = Command.parse(request.command);
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ProtocolDispatcher.java]: handle for UNSUPPORTED command

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ProtocolDispatcher.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import org.eclipse.jdt.ls.debug.DebugException;
+import org.eclipse.jdt.ls.debug.IDebugSession;
+import org.eclipse.jdt.ls.debug.adapter.Events.DebugEvent;
+import org.eclipse.jdt.ls.debug.adapter.Messages.Request;
+import org.eclipse.jdt.ls.debug.adapter.Messages.Response;
+import org.eclipse.jdt.ls.debug.adapter.Requests.Arguments;
+import org.eclipse.jdt.ls.debug.adapter.Requests.Command;
+import org.eclipse.jdt.ls.debug.adapter.handler.InitializeRequestHandler;
+
+public class ProtocolDispatcher implements IDebugAdapter, IDebugAdapterContext {
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ProtocolDispatcher.java]: split IDebugAdapterContext a new instance.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 3]
```diff
@@ -99,249 +120,236 @@
...
         this.eventSubscriptions = new ArrayList<>();
-        this.context = context;
+        this.providerContext = providerContext;
         this.variableRequestHandler = new VariableRequestHandler(VariableFormatterFactory.createVariableFormatter());
+        this.debugContext = new DebugAdapterContext(this);
+        this.requestHandlers = new HashMap<>();
+        initialize();
+    }
+
+    private void initialize() {
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: move private methods to the end of all public methods

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 4]
```diff
@@ -610,8 +598,8 @@ private void launchDebugSession(Requests.LaunchArguments arguments) throws Debug
             });
             debuggeeConsole.start();
         } catch (IOException | IllegalConnectorArgumentsException | VMStartException e) {
-            Logger.logException("Launching debuggee vm exception", e);
-            throw new DebugException("Launching debuggee vm exception \"" + e.getMessage() + "\"", e);
+            Logger.logException("Failed to launch debuggee vm. Reason: \"" + e.getMessage() + "\"", e);
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: String.format

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 5]
```diff
@@ -610,8 +598,8 @@ private void launchDebugSession(Requests.LaunchArguments arguments) throws Debug
...
+            throw new DebugException("Failed to launch debuggee vm. Reason: \"" + e.getMessage() + "\"", e);
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: don't construct the error message twice as same as above line

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ErrorCodes.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.adapter;
+
+public final class ErrorCodes {
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ErrorCodes.java]: enum error codes.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/AdapterUtils.java [hunk 7]
```diff
@@ -67,4 +73,96 @@ public static String parseEnclosingType(String fullyQualifiedName) {
...
+     *
+     * @param line
+     *           the line number from the source platform
+     * @param sourceLinesStartAt1
+     *           the source platform's line starts at 1 or not
+     * @param targetLinesStartAt1
+     *           the target platform's line starts at 1 or not
+     * @return the new line number
+     */
+    public static int formatLineNumber(int line, boolean sourceLinesStartAt1, boolean targetLinesStartAt1) {
```
Comment (ansyral) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/AdapterUtils.java]: >formatLineNumber [](start = 22, length = 16)

minor, method name doesn't convey what the method does without reading the comments

Comment (ansyral) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/AdapterUtils.java]: the same to formatPath

---
In reply to: [135172222](https://github.com/vscjavaci/eclipse.jdt.ls/pull/48#discussion_r135172222) [](ancestors = 135172222)

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ErrorCode.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.adapter;
+
+public enum ErrorCode {
+    UNKNOWN_FAILURE,
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ErrorCode.java]: Please assign values explicitly to each member.

**Review by akaroml — DISMISSED** [review]

**Review by akaroml — APPROVED** [review]

**Pull request merged**